targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/trng_api.c

```diff
@@ -0,0 +1,96 @@
+ *     that would cause the software to become subject to terms and conditions
+ *     which differ from those listed here.
+ *   - Neither the name of Analog Devices, Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *   - The use of this software may or may not infringe the patent rights of one
+ *     or more patent holders.  This license does not release you from the
+ *     requirement that you obtain separate licenses from these patent holders
+ *     to use this software.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES, INC. AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-
+ * INFRINGEMENT, TITLE, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ANALOG DEVICES, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, DAMAGES ARISING OUT OF
+ * CLAIMS OF INTELLECTUAL PROPERTY RIGHTS INFRINGEMENT; PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#if defined(DEVICE_TRNG)
+
+#include <stdlib.h>
+#include <adi_rng.h>
+#include <adi_pwr.h>
+#include "cmsis.h"
+#include "trng_api.h"
+
+// Sampling counter values
+// Prescaler: 0 - 10
+// LenReload: 0 - 4095
+#define TRNG_CNT_VAL	4095
+#define TRNG_PRESCALER 	2
+
+static ADI_RNG_HANDLE     RNGhDevice;                     /* Memory to handle CRC Device */
+static uint32_t           RngDevMem[(ADI_RNG_MEMORY_SIZE + 3)/4]; /* Data buffers for Random numbers */
+
+void trng_init(trng_t *obj)
+{
+    (void)obj;
+    adi_rng_Open(0,RngDevMem,sizeof(RngDevMem),&RNGhDevice);
+
+    // Set sample length for the H/W RN accumulator
+    adi_rng_SetSampleLen(RNGhDevice, TRNG_PRESCALER, TRNG_CNT_VAL);
+
+    // Enable the RNG
+    adi_rng_Enable(RNGhDevice, true);
+}
+
+void trng_free(trng_t *obj)
+{
+    (void)obj;
+    adi_rng_Enable(RNGhDevice, false);
+    adi_rng_Close(RNGhDevice);
+}
+
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    (void)obj;
+    bool bRNGRdy;
+    uint32_t nRandomNum, i;
+
+    for (i = 0; i < length; ) {
+        // wait for the RNG ready to give a random number
+        adi_rng_GetRdyStatus(RNGhDevice, &bRNGRdy);
+        if (bRNGRdy) {
+            // read the random number
+            adi_rng_GetRngData(RNGhDevice, &nRandomNum);
+            output[i++] = (uint8_t) nRandomNum;
```

Comment:
`output` is a buffer of `uint8_t`. `nRandomNum` is `uint32_t`. ALthough you downcast to uint8_t, I assume that `adi_rng_GetRngData` collects 32 bits of entropy, but you take only 8 bits each time. You loose data this way, and it is likely that the output data will not have full entropy. 
It is better to take the full random word and fill it in `output` buffer, as long as it fits. if `length` doesn't divide with `sizeof(uint32_t)`. the last collection should only take part of the `nRandomNum` value